File: api/internal/handler/cli_tokens.go
package handler

import (
	"log/slog"
	"net/http"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgtype"

	"github.com/vtmocanu/uzi/api/internal/apitypes"
	"github.com/vtmocanu/uzi/api/internal/clitoken"
	"github.com/vtmocanu/uzi/api/internal/httpx"
	mw "github.com/vtmocanu/uzi/api/internal/middleware"
	"github.com/vtmocanu/uzi/api/internal/store"
	"github.com/vtmocanu/uzi/api/internal/termsafe"
)

// maxCLITokenNameBytes bounds the human label on a CLI token, matching the worker
// name cap.
const maxCLITokenNameBytes = 200

// cliTokenTTL is the server-set lifetime for every bounded CLI token (Decision 8):
// 90 days. The expiry matrix bounds by SCOPE first (a uza_ is always 90 days on
// either acquisition path) and by ACQUISITION PATH for user tokens — a webui-minted
// uzc_ (this endpoint) never expires, because the agent/CI path is the one footgun
// a silent mid-pipeline death cannot absorb. The browser-brokered uzc_ (M5) is
// bounded to 90 days. The client never proposes a lifetime; the server sets it.
const cliTokenTTL = 90 * 24 * time.Hour

// cliTokenDTO is the metadata view of a CLI token. It lives HERE in the handler
// package, not in apitypes: no Go CLI verb decodes this endpoint (it is cookie-only;
// the SPA consumes it), so it is not a binary-hygiene type. The token VALUE is never
// a field — it is shown once at mint and never stored, so it can never be listed.
//
// token_prefix + last_used_at + last_used_ip are the ENTIRE forensic surface
// (Risk 8): there is no per-request audit log and a password change does not revoke
// CLI tokens, so "which token is this, and was it used by someone who isn't me?" is
// answerable only from these three. They are not optional columns.
type cliTokenDTO struct {
	ID          string     `json:"id"`
	Name        string     `json:"name"`
	TokenPrefix string     `json:"token_prefix"`
	Scope       string     `json:"scope"`
	Revoked     bool       `json:"revoked"`
	CreatedAt   time.Time  `json:"created_at"`
	LastUsedAt  *time.Time `json:"last_used_at"`
	LastUsedIP  *string    `json:"last_used_ip"`
	ExpiresAt   *time.Time `json:"expires_at"`
}

func cliTokenToDTO(t store.CliToken) cliTokenDTO {
	dto := cliTokenDTO{
		ID:          t.ID.String(),
		Name:        t.Name,
		TokenPrefix: t.TokenPrefix,
		Scope:       t.Scope,
		Revoked:     t.Revoked,
		CreatedAt:   t.CreatedAt.Time,
		LastUsedAt:  timePtr(t.LastUsedAt.Valid, t.LastUsedAt.Time),
		ExpiresAt:   timePtr(t.ExpiresAt.Valid, t.ExpiresAt.Time),
	}
	if t.LastUsedIp != nil {
		s := t.LastUsedIp.String()
		dto.LastUsedIP = &s
	}
	return dto
}

// ListCLITokens returns the caller's CLI tokens (metadata only, never the value).
func (h *Handler) ListCLITokens(w http.ResponseWriter, r *http.Request) {
	user, ok := mw.UserFromContext(r.Context())
	if !ok {
		httpx.Error(w, http.StatusUnauthorized, "authentication required")
		return
	}
	rows, err := h.q.ListCLITokens(r.Context(), user.ID)
	if err != nil {
		slog.Error("list cli tokens", "error", err)
		httpx.Error(w, http.StatusInternalServerError, "internal error")
		return
	}
	out := make([]cliTokenDTO, 0, len(rows))
	for _, t := range rows {
		out = append(out, cliTokenToDTO(t))
	}
	httpx.JSON(w, http.StatusOK, map[string]any{"tokens": out})
}

// CreateCLIToken mints a CLI token and returns its plaintext value exactly once
// (only the sha256 is stored), like CreateWorker. The server sets the expiry per
// the matrix; the client never proposes one. admin_ro is never the default and only
// an admin may mint it.
func (h *Handler) CreateCLIToken(w http.ResponseWriter, r *http.Request) {
	user, ok := mw.UserFromContext(r.Context())
	if !ok {
		httpx.Error(w, http.StatusUnauthorized, "authentication required")
		return
	}
	var req struct {
		Name  string `json:"name"`
		Scope string `json:"scope"`
	}
	if err := httpx.DecodeJSON(r, &req); err != nil {
		httpx.Error(w, http.StatusBadRequest, "invalid request body")
		return
	}
	name := strings.TrimSpace(req.Name)
	if name == "" || len(name) > maxCLITokenNameBytes {
		httpx.Error(w, http.StatusBadRequest, "name must be non-empty and at most 200 characters")
		return
	}
	// #169, same rule and same reason as CreateWorker: `uzi admin cli-tokens` prints this
	// name beside another user's owner_email. termsafe.Validate rejects exactly what the
	// CLI's CellText would strip, so a stored name round-trips to display unchanged.
	if err := termsafe.Validate("name", name); err != nil {
		httpx.Error(w, http.StatusBadRequest, err.Error())
		return
	}
	scope := req.Scope
	if scope == "" {
		scope = clitoken.ScopeUser
	}
	switch scope {
	case clitoken.ScopeUser:
		// default; capped to the owner's own authority.
	case clitoken.ScopeAdminRO:
		// admin_ro reads the whole factory, so only an admin may mint it — and it is
		// never the default (Risk 8 least-privilege). Resolved live from the row.
		if !user.IsAdmin {
			httpx.Error(w, http.StatusForbidden, "admin access required to mint an admin-scoped token")
			return
		}
	default:
		httpx.Error(w, http.StatusBadRequest, "scope must be 'user' or 'admin_ro'")
		return
	}

	token, hash, prefix, err := clitoken.Generate(scope)
	if err != nil {
		slog.Error("generate cli token", "error", err)
		httpx.Error(w, http.StatusInternalServerError, "internal error")
		return
	}

	// Expiry matrix (M2 = the webui static mint path): a user-scope token never
	// expires (NULL); an admin_ro token is bounded to 90 days. The browser-brokered
	// user token's 90-day bound is M5's, at its own mint site.
	var expires pgtype.Timestamptz
	if scope == clitoken.ScopeAdminRO {
		expires = pgtype.Timestamptz{Time: time.Now().Add(cliTokenTTL), Valid: true}
	}

	row, err := h.q.CreateCLIToken(r.Context(), store.CreateCLITokenParams{
		UserID:      user.ID,
		Name:        name,
		TokenHash:   hash,
		TokenPrefix: prefix,
		Scope:       scope,
		ExpiresAt:   expires,
	})
	if err != nil {
		slog.Error("create cli token", "error", err)
		httpx.Error(w, http.StatusInternalServerError, "internal error")
		return
	}

	if scope == clitoken.ScopeAdminRO {
		// The only detection breadcrumb for a factory-wide-read credential's mint
		// (Risk 8): no per-request audit log exists, so record the mint itself.
		slog.Info("admin_ro cli token minted", "user_id", user.ID, "token_id", row.ID, "token_prefix", prefix)
	}

	httpx.JSON(w, http.StatusCreated, map[string]any{
		"token":     token,
		"cli_token": cliTokenToDTO(row),
	})
}

// RevokeCLIToken soft-deletes one of the caller's CLI tokens. Owner-scoped: a
// foreign or unknown id is a 404, never a cross-user revoke.
func (h *Handler) RevokeCLIToken(w http.ResponseWriter, r *http.Request) {
	user, ok := mw.UserFromContext(r.Context())
	if !ok {
		httpx.Error(w, http.StatusUnauthorized, "authentication required")
		return
	}
	id, ok := httpx.PathUUID(w, r, "id", "token")
	if !ok {
		return
	}
	n, err := h.q.RevokeCLIToken(r.Context(), store.RevokeCLITokenParams{ID: id, UserID: user.ID})
	if err != nil {
		slog.Error("revoke cli token", "error", err)
		httpx.Error(w, http.StatusInternalServerError, "internal error")
		return
	}
	if n == 0 {
		httpx.Error(w, http.StatusNotFound, "token not found")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// RevokeAllCLITokens revokes every un-revoked token of the caller — the panic
// button for a lost laptop (Decision 19). Idempotent: a second call is a no-op that
// still returns 204. Scoped to the caller, so it never touches another user's tokens.
func (h *Handler) RevokeAllCLITokens(w http.ResponseWriter, r *http.Request) {
	user, ok := mw.UserFromContext(r.Context())
	if !ok {
		httpx.Error(w, http.StatusUnauthorized, "authentication required")
		return
	}
	if err := h.q.RevokeAllCLITokens(r.Context(), user.ID); err != nil {
		slog.Error("revoke all cli tokens", "error", err)
		httpx.Error(w, http.StatusInternalServerError, "internal error")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// AdminListCLITokens returns every CLI token in the factory with its owner, for the
// admin standing-credential inventory. Read-only: there is deliberately no admin
// revoke here (see the route comment).
//
// Admin-gating is the route group's (RequireUser + RequireAdminRO), which resolves
// admin-ness LIVE from the user row and has already masked any non-admin_ro token to
// IsAdmin=false — so this handler must not re-check scope, exactly as the other
// admin reads do not. One mechanism, one place.
func (h *Handler) AdminListCLITokens(w http.ResponseWriter, r *http.Request) {
	rows, err := h.q.ListAllCLITokensForAdmin(r.Context())
	if err != nil {
		slog.Error("admin list cli tokens", "error", err)
		httpx.Error(w, http.StatusInternalServerError, "internal error")
		return
	}
	out := make([]apitypes.AdminCLITokenDTO, 0, len(rows))
	for _, t := range rows {
		dto := apitypes.AdminCLITokenDTO{
			ID:          t.ID.String(),
			UserID:      t.UserID.String(),
			OwnerEmail:  t.OwnerEmail,
			Name:        t.Name,
			TokenPrefix: t.TokenPrefix,
			Scope:       t.Scope,
			Revoked:     t.Revoked,
			CreatedAt:   t.CreatedAt.Time,
			LastUsedAt:  timePtr(t.LastUsedAt.Valid, t.LastUsedAt.Time),
			ExpiresAt:   timePtr(t.ExpiresAt.Valid, t.ExpiresAt.Time),
		}
		if t.LastUsedIp != nil {
			s := t.LastUsedIp.String()
			dto.LastUsedIP = &s
		}
		out = append(out, dto)
	}
	httpx.JSON(w, http.StatusOK, map[string]any{"tokens": out})
}

File: web/src/mocks/mockApi/cliTokens.ts
import type { CliAuthRequestMeta, CliToken, CliTokenScope } from "../../lib/api";
import { ApiError } from "../../lib/apiError";
import { MOCK_CLI_AUTH_REQUEST_ID, mockCliAuthRequest, mockCliTokens } from "../data";
import { delay, requireSession } from "./shared";

// Tokens are owner-attributed (user_id) so every read/write is scoped to the
// session user, mirroring the real endpoints (`WHERE user_id=$1`). user_id is
// mock-internal — stripped before responding, since the wire CliToken has none.
type OwnedCliToken = CliToken & { user_id: string };
let cliTokens: OwnedCliToken[] = mockCliTokens.map((t) => ({ ...t }));
let cliTokenCounter = 0;
const stripOwner = ({ user_id: _user_id, ...t }: OwnedCliToken): CliToken => t;

// The seeded consent request; approve/deny flip its status in place.
const cliAuthRequests = new Map<string, CliAuthRequestMeta & { user_code: string }>([
  [MOCK_CLI_AUTH_REQUEST_ID, { ...mockCliAuthRequest }],
]);

// crockford32 excludes I, L, O, U. normalizeMockUserCode folds a human-typed code
// to the canonical stored form, mirroring the server (uppercase; hyphens/spaces
// dropped; O→0, I/L→1) so a careful-but-imperfect read still matches.
const CROCKFORD32 = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";
function normalizeMockUserCode(s: string): string {
  let out = "";
  for (const r of s.toUpperCase()) {
    if (r === "O") out += "0";
    else if (r === "I" || r === "L") out += "1";
    else if (CROCKFORD32.includes(r)) out += r;
  }
  return out;
}

export const cliTokensApi = {
  // ── CLI tokens (PRD #64 M6) ────────────────────────────────────────────────
  // Mirrors the cookie-only CRUD: list carries no value, mint returns the
  // plaintext once, admin_ro is admin-only, revoked rows stay (the incident
  // trail), and revoke-all is idempotent.
  listCliTokens: async () => {
    const me = requireSession();
    // Only the caller's own tokens — the real endpoint filters `WHERE user_id=$1`,
    // so as a non-admin persona you must not see the admin's tokens.
    return delay({ tokens: cliTokens.filter((t) => t.user_id === me.id).map(stripOwner) });
  },
  createCliToken: async (name: string, scope: CliTokenScope) => {
    const me = requireSession();
    const trimmed = name.trim();
    if (trimmed === "") throw new ApiError(400, "name must be non-empty and at most 200 characters");
    if (scope !== "user" && scope !== "admin_ro") throw new ApiError(400, "scope must be 'user' or 'admin_ro'");
    if (scope === "admin_ro" && !me.is_admin) {
      throw new ApiError(403, "admin access required to mint an admin-scoped token");
    }
    const cls = scope === "admin_ro" ? "uza" : "uzc";
    const body = Array.from(crypto.getRandomValues(new Uint8Array(24)), (b) => b.toString(16).padStart(2, "0")).join("");
    const now = new Date().toISOString();
    const row: OwnedCliToken = {
      id: `cli-new-${++cliTokenCounter}`,
      user_id: me.id,
      name: trimmed,
      token_prefix: `${cls}_${body.slice(0, 4)}`,
      scope,
      revoked: false,
      created_at: now,
      last_used_at: null,
      last_used_ip: null,
      // Expiry matrix (static mint path): a user token never expires; an admin_ro
      // token is bounded to 90 days.
      expires_at: scope === "admin_ro" ? new Date(Date.now() + 90 * 86_400_000).toISOString() : null,
    };
    cliTokens = [row, ...cliTokens];
    return delay({ token: `${cls}_${body}`, cli_token: stripOwner(row) }, 200);
  },
  revokeCliToken: async (id: string) => {
    const me = requireSession();
    // Owner-scoped: a foreign id is a 404, exactly like the server.
    const t = cliTokens.find((x) => x.id === id && x.user_id === me.id && !x.revoked);
    if (!t) throw new ApiError(404, "token not found");
    t.revoked = true;
    return delay(null);
  },
  revokeAllCliTokens: async () => {
    const me = requireSession();
    // Only the caller's tokens, mirroring `WHERE user_id=$1`.
    cliTokens = cliTokens.map((t) => (t.user_id === me.id ? { ...t, revoked: true } : t));
    return delay(null);
  },

  // ── CLI browser-login consent flow (PRD #64 M6) ────────────────────────────
  // getCliAuthRequest never returns the user_code (the human types it from their
  // terminal — the anti-phishing property). approve validates the typed code.
  getCliAuthRequest: async (id: string) => {
    requireSession();
    const req = cliAuthRequests.get(id);
    if (!req) throw new ApiError(404, "request not found");
    const status =
      req.status === "pending" && Date.parse(req.expires_at) <= Date.now() ? "expired" : req.status;
    return delay({ client_desc: req.client_desc, status, expires_at: req.expires_at });
  },
  approveCliAuth: async (requestId: string, userCode: string, scope: CliTokenScope) => {
    const me = requireSession();
    if (scope !== "user" && scope !== "admin_ro") throw new ApiError(400, "scope must be 'user' or 'admin_ro'");
    if (scope === "admin_ro" && !me.is_admin) {
      throw new ApiError(403, "admin access required to approve an admin-scoped login");
    }
    const req = cliAuthRequests.get(requestId);
    if (!req) throw new ApiError(404, "request not found");
    if (Date.parse(req.expires_at) <= Date.now()) throw new ApiError(410, "request expired");
    if (req.status !== "pending") throw new ApiError(409, "request is no longer pending");
    if (normalizeMockUserCode(userCode) !== req.user_code) {
      throw new ApiError(400, "the code you entered does not match");
    }
    req.status = "approved";
    return delay({ status: "approved" });
  },
  denyCliAuth: async (requestId: string) => {
    requireSession();
    const req = cliAuthRequests.get(requestId);
    if (!req) throw new ApiError(404, "request not found");
    if (req.status !== "pending" || Date.parse(req.expires_at) <= Date.now()) {
      throw new ApiError(409, "request is no longer pending");
    }
    req.status = "denied";
    return delay({ status: "denied" });
  },
};
